File: sdk/python/polar/customer_portal/customer_meters.py
from __future__ import annotations

import builtins

from polar.base import AsyncServiceBase, SyncServiceBase, parse_response
from polar.errors import (
    HTTPValidationError,
    ResourceNotFound,
)
from polar.literals import (
    CustomerCustomerMeterSortProperty,
)
from polar.outputs import (
    CustomerCustomerMeter,
    ListResourceCustomerCustomerMeter,
)


class CustomerMetersSync(SyncServiceBase):
    def list(
        self,
        *,
        meter_id: str | builtins.list[str] | None = None,
        query: str | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[CustomerCustomerMeterSortProperty] | None = [
            "-modified_at"
        ],
    ) -> ListResourceCustomerCustomerMeter:
        """
        List meters of the authenticated customer.

        **Scopes**: `customer_portal:read` `customer_portal:write`

        Args:
            meter_id: Filter by meter ID.
            query: Filter by meter name.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/meters/",
            path_params={},
            query_params={
                "meter_id": meter_id,
                "query": query,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(
            response, ListResourceCustomerCustomerMeter, method_errors
        )

    def get(
        self,
        id: str,
    ) -> CustomerCustomerMeter:
        """
        Get a meter by ID for the authenticated customer.

        **Scopes**: `customer_portal:read` `customer_portal:write`

        Args:
            id: The customer meter ID.

        Raises:
            ResourceNotFound: Customer meter not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/meters/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, CustomerCustomerMeter, method_errors)


class CustomerMetersAsync(AsyncServiceBase):
    async def list(
        self,
        *,
        meter_id: str | builtins.list[str] | None = None,
        query: str | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[CustomerCustomerMeterSortProperty] | None = [
            "-modified_at"
        ],
    ) -> ListResourceCustomerCustomerMeter:
        """
        List meters of the authenticated customer.

        **Scopes**: `customer_portal:read` `customer_portal:write`

        Args:
            meter_id: Filter by meter ID.
            query: Filter by meter name.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/meters/",
            path_params={},
            query_params={
                "meter_id": meter_id,
                "query": query,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(
            response, ListResourceCustomerCustomerMeter, method_errors
        )

    async def get(
        self,
        id: str,
    ) -> CustomerCustomerMeter:
        """
        Get a meter by ID for the authenticated customer.

        **Scopes**: `customer_portal:read` `customer_portal:write`

        Args:
            id: The customer meter ID.

        Raises:
            ResourceNotFound: Customer meter not found.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/meters/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, CustomerCustomerMeter, method_errors)

File: sdk/typescript/src/services/customer_portal/customer_meters.ts
import { ClientBase } from "../../base";
import type {
  CustomerCustomerMeter,
  ListResourceCustomerCustomerMeter,
} from "../../models/outputs";
import type { CustomerCustomerMeterSortProperty } from "../../models/literals";
import { HTTPValidationError, ResourceNotFound } from "../../errors";

export const listCustomerMeters = (client: ClientBase) => {
  /**
   * List meters of the authenticated customer.
   *
   * **Scopes**: `customer_portal:read` `customer_portal:write`
   *
   * @param query - Query parameters
   * @returns {ListResourceCustomerCustomerMeter}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query?: {
    meter_id?: string | string[] | null;
    query?: string | null;
    page?: number;
    limit?: number;
    sorting?: CustomerCustomerMeterSortProperty[] | null;
  }): Promise<ListResourceCustomerCustomerMeter> => {
    const pathParams = {};
    const queryParams = {
      meter_id: query?.meter_id,
      query: query?.query,
      page: query?.page || 1,
      limit: query?.limit || 10,
      sorting: query?.sorting || ["-modified_at"],
    };
    const request = client.buildRequest(
      "GET",
      "/v1/customer-portal/meters/",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<ListResourceCustomerCustomerMeter>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const getCustomerMeters = (client: ClientBase) => {
  /**
   * Get a meter by ID for the authenticated customer.
   *
   * **Scopes**: `customer_portal:read` `customer_portal:write`
   *
   * @param id - The customer meter ID.
   * @returns {CustomerCustomerMeter}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Customer meter not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<CustomerCustomerMeter> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/customer-portal/meters/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CustomerCustomerMeter>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};

export function createCustomerMetersService(client: ClientBase) {
  return {
    list: listCustomerMeters(client),
    get: getCustomerMeters(client),
  };
}

export type CustomerMeters = ReturnType<typeof createCustomerMetersService>;
